### PR TITLE
Automate Deployment of Caasp4 using terraform

### PR DESCRIPTION
**Pull request by mbelur**
This playbook provides an automated deployment mechanism
to deploy caasp v4 using terraform on SOC openstack deployment.

This playbook currently runs against the ardana flavored version of
SOC.

You can provide various optional parameters to the playbook:
Some examples:
num_workers to define the number of workers in the cluster.
	- Default value is 1.
num_masters to define the number of masters in the cluster.
	- Default value is 1.

It uses octavia provider for SOC 9 and haproxy provider for SOC 8
deployments.

### CODE REVIEW

**Comment by mbelur** [comment]
Example runs:

1. https://ci.suse.de/job/openstack-ardana-caaspv4/23/
	SOC 9 Virtual Cloud Environment with default num_workers=1 and num_masters=1

2. https://ci.suse.de/job/openstack-ardana-caaspv4/26/
	SOC 9 Baremetal QE Environment with extra_params set to:
	num_workers=3
	num_masters=3

3. https://ci.suse.de/job/openstack-ardana/579/
	SOC 8 Virtual Cloud Environment with_caaspv4 enabled as part of the deployment

4. https://ci.suse.de/job/openstack-ardana/580/
	SOC 8 Baremetal Environment with_caaspv4 enabled as part of the deployment 

5. https://ci.suse.de/job/openstack-ardana/581/
	SOC 9 Virtual Cloud Environment with_caaspv4 enabled as part of the deployment

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+
+  - name: Upload SLES15SP1 image to glance
+    shell: |
+      source "{{ service_osrc }}"
+      if ! openstack image list | grep "{{ image_name }}"; then openstack image create --public --disk-format qcow2 --container-format bare --file "{{ image_path }}" "{{ image_name }}"; fi
+    args:
+      executable: /bin/bash
+
+  - name: Create Management Machine
+    shell: |
```
Comment (jsuchome) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: I'm not sure what's the usual way. But as this is just a large bash script, wouldn't it be better to save it to the file and execute the script from ansible?

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+  - name: Managment Machine | Resources names with timestamp
+    # resources named with timestamp
+    set_fact:
+      stack_name: "{{stack_name + '-' + timestamp}}"
+      internal_net: "{{internal_net + '-' + timestamp}}"
+      internal_subnet: "{{internal_subnet + '-' + timestamp}}"
+
+  - name: Management Machine | Update resolv.conf
+    shell:
+      echo "nameserver 10.84.2.20" | sudo tee -a /etc/resolv.conf
```
Comment (jsuchome) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: Could you put `10.84.2.20` into a variable?

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+      name: "SUSE-CaaSP-Management"
+      state: present
+      type: pattern
+    become: yes
+
+  - name: Management Machine | Copy caasp
+    shell: |
+      set -e
+      mkdir -p ~/caasp/deployment/
+      cp -r /usr/share/caasp/terraform/openstack/ ~/caasp/deployment/
```
Comment (jsuchome) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: Why is there `caasp_openstack_path` variable, but here you are using `~/caasp/deployment/` directly?

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+      terraform init | tee -a terraform_init.log
+    register:
+      terraform_init_results
+    failed_when: "not ('Terraform has been successfully initialized!' in terraform_init_results.stdout)"
+
+  - name: Management Machine |  Run terraform plan
+    shell: |
+      set -e
+      eval `ssh-agent -s`
+      ssh-add
```
Comment (jsuchome) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: Could the host machine be set up so it starts ssh-agent automatically, so you do not need to invoke it in each terraform call?

**Comment by jsuchome** [comment]
You might actually reuse some parts of the decommissioned socok8s code, see e.g. here:

https://github.com/SUSE-Cloud/socok8s/tree/archive-add_deploy_process_thing-WIP/playbooks/roles/caasp4

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+    stack_name: "auto-caasp"
+    subnet_cidr: "172.28.0.0/24"
+    terraform_destroy: false
+    tfvars_path: "{{ caasp_openstack_path }}/terraform.tfvars"
+    timestamp: ""
+    service_caasp_osrc: "service_caasp.osrc"
+    user: "sles"
+    worker_size: "m1.medium"
+  tasks:
+  - name: Use python3
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: Do you really need this task? wouldn't setting this alongside the other playbooks variables work?

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+      path: "{{ registration_auto_vars_path }}"
+      regexp: '^#?rmt_server_name ='
+      line: 'rmt_server_name = "rmt.scc.suse.de"'
+
+  - name: Management Machine | Update terraform.tfvars
+    lineinfile:
+      dest: "{{ item.dest }}"
+      regexp: "{{ item.regexp }}"
+      line: "{{ item.line }}"
+    with_items:
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: you could use `loop` here as `with_` is deprecated

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+    when: item|trim is regex("skuba node bootstrap .*$")
+    with_items:
+    - "{{ terraform_apply_results.stdout_lines }}"
+
+  - name: Management Machine |  Parse skuba join commands
+    set_fact:
+      skuba_join_commands: "{{ skuba_join_commands | default([])}} +  {{ [item | regex_replace('(^.*CMD ]  )', '')] }}"
+    when: item|trim is regex("skuba node join .*$")
+    with_items:
+    - "{{ terraform_apply_results.stdout_lines }}"
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: same here `loop`

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+      ssh -o StrictHostKeyChecking=no sles@{{item}} sudo update-ca-certificates
+    with_items: "{{ master_ips.value.values() | list }}"
+
+  - name: Management Machine | Copy certificates to each of the worker nodes in the cluster
+    shell: |
+      set -e
+      scp -o StrictHostKeyChecking=no {{ ca_path }} sles@{{item}}:~/ca-bundle.pem
+      ssh -o StrictHostKeyChecking=no sles@{{item}}  sudo cp ~/ca-bundle.pem {{ ca_path }}
+      ssh -o StrictHostKeyChecking=no sles@{{item}} sudo update-ca-certificates
+    with_items: "{{ worker_ips.value.values() | list }}"
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: same here `loop`

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+      skuba_mv_openstack_conf_command_results
+
+  - name: Management Machine | Run skuba bootstrap
+    shell: |
+      set -e
+      eval `ssh-agent -s`
+      ssh-add
+      cd {{ caasp_openstack_path }}/{{stack_name}}-cluster
+      {{ item }} | tee -a skuba_bootstrap.log
+    with_items: "{{ skuba_boostrap_commands }}"
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: same here `loop`

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+      skuba_bootstrap_commands_results
+    failed_when: " not ('[bootstrap] successfully bootstrapped core add-ons' in skuba_bootstrap_commands_results.stdout)"
+
+  - name: Set expected total nodes in the cluster
+    set_fact:
+      total: "{{ num_masters | int + num_workers | int}}"
+
+  - name: Set expected total nodes joining the cluster after bootstrap
+    set_fact:
+      total_join: "{{ num_masters | int + num_workers | int - 1}}"
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: I think you missed a space before `}}`

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+      ssh-add
+      cd {{ caasp_openstack_path }}/{{stack_name}}-cluster
+      {{ item }} | tee -a skuba_join.log
+    with_items: "{{ skuba_join_commands }}"
+    register:
+      skuba_join_commands_results
+
+  - name: Management Machine | track the count of the nodes that join the cluster successfully
+    set_fact:
+      count: "{{ count | default(0) | int + 1}}"
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: same here, missing space before `}}`

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+
+  - name: Management Machine | Fail when the expected number of nodes do not join the cluster
+    fail:
+      msg: "Not all expected nodes have successfully joined the cluster"
+    when: "count | int != total_join | int"
+
+  - name: Management Machine | Run skuba cluster status to validate all the nodes are in the cluster in Ready state
+    shell: |
+      set -e
+      cd {{ caasp_openstack_path }}/{{stack_name}}-cluster
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: keeping the same pattern for variables, here `stack_name` does not have white spaces before/after `{{ }}`

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+    octavia_enabled: true
+    os_ca_path: "/etc/ssl/ca-bundle.pem"
+    service_osrc: "{{ soc_home }}/service.osrc"
+    ssh_opts: '-o BatchMode=yes -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o LogLevel=error -o ConnectTimeout=30'
+    service_caasp_osrc: "service_caasp.osrc"
+  tasks:
+
+  #for SOC 8: Use the regular Neutron LBaaS API
+  #conditional check ansible_distribution_release == "3" means SLES12SP3 deployer
+  - set_fact:
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: Do you really need that task? What about setting this as variable, like:
`
octavia_enabled: "{{ not (ansible_distribution_release == '3') }}"
`

File: scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+  - name: Get public vip
+    shell: grep  'public-KEY-API-extapi' /etc/hosts | awk {'print $1'}
+    register: public_vip
+
+  - name: Update service_caasp.osrc
+    lineinfile:
+      dest: "{{ item.dest }}"
+      regexp: "{{ item.regexp }}"
+      line: "{{ item.line }}"
+    with_items:
```
Comment (flaviodsr) [inline scripts/jenkins/cloud/ansible/deploy-caasp-v4-terraform.yml]: you could use `loop` here as `with_` is deprecated

**Pull request merged**